packages/core/contracts/bot-helpers/uniswap-broker/UniswapBrokerV3.sol

```diff
@@ -0,0 +1,629 @@
+        while (true) {
+            // Compute the next initialized tick. We only need to traverse initialized ticks as uninitialized ticks
+            // have the same liquidity as the previous tick.
+            StepComputations memory step;
+            step.sqrtPriceStartX96 = state.sqrtPriceX96;
+            (step.tickNext, step.initialized) = TickBitmap.nextInitializedTickWithinOneWord(
+                pool,
+                state.tick,
+                pool.tickSpacing(),
+                zeroForOne
+            );
+
+            // Double check we are not over or underflow the ticks.
+            if (step.tickNext < TickMath.MIN_TICK) step.tickNext = TickMath.MIN_TICK;
+            else if (step.tickNext > TickMath.MAX_TICK) step.tickNext = TickMath.MAX_TICK;
+
+            // Find the price at the next tick. Between the current state.sqrtPriceX96 and the nextTickPriceX96 we
+            // can find how much of the sold token is needed to sufficiently move the market over the interval.
+
+            uint160 nextTickPriceX96 = TickMath.getSqrtRatioAtTick(step.tickNext);
+            uint256 inputAmountForStep;
+
+            // If zeroForOne is true, then we are moving the price UP. In this case we need to ensure that if the next
+            // tick price is more than the target price, we set the set the the next step price to the target price. This
+            // ensures that the price does not undershoot when the next tick is the last tick. Else, traverse the whole tick.
+            if (zeroForOne) {
+                step.sqrtPriceNextX96 = nextTickPriceX96 > sqrtRatioTargetX96 ? sqrtRatioTargetX96 : nextTickPriceX96;
+                inputAmountForStep = SqrtPriceMath.getAmount0Delta( // As we are trading token0 for token1, calculate the token0 input.
+                    step.sqrtPriceStartX96,
+                    step.sqrtPriceNextX96,
+                    state.liquidity,
+                    false
+                );
+                // Else, if zeroForOne is false, then we are moving the price UP. In this case we need to ensure that we
+                // don't overshoot the price on the next step.
+            } else {
+                step.sqrtPriceNextX96 = nextTickPriceX96 > sqrtRatioTargetX96 ? nextTickPriceX96 : sqrtRatioTargetX96;
+                inputAmountForStep = SqrtPriceMath.getAmount1Delta( // As we are trading token1 for token0, calculate the token1 input.
+                    step.sqrtPriceStartX96,
+                    step.sqrtPriceNextX96,
+                    state.liquidity,
+                    false
+                );
+            }
+
+            // Add amount for this step to the total required input.
+            state.requiredInputAmount = state.requiredInputAmount.add(inputAmountForStep);
+
+            // If we have hit(or exceeded) our target price in the associate direction, then stop.
+            if (zeroForOne && state.sqrtPriceX96 <= sqrtRatioTargetX96) break;
```

Comment:
Why don't you check these earlier at the beginning of the loop? Otherwise you might do some unneccessary computations  before you `break`

Reply:
good question. the reason is that on the _first_ loop you need to add in the initial amount _before_ updating the liquidity. Doing it in this ordering means that on the _initial_ liquidity, pulled in from `pool.liquidity`, we calculate the initial `requiredInputAmount` traversed before updating the liquditiy.

Reply:
Ah interesting, you could try using a `do...while`? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/do...while

Reply:
that would work, ye. The way this is set up though we actually dont want to break until after the `state.requiredInputAmount` has been updated  with the next `inputAmountForStep`.

i.e if we used a `do...while` the loop would exit out as soon as the `state.sqrtPriceX96` hits the `sqrtRatioTargetX96` and it would not increment the `requiredInputAmount`.

More over, there are two different break conditions depending on if `zeroForOne` is true (or not). This would be pretty messy to deal with in the `do...while` syntax. For these reason I think I want to leave it as it is now.